compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/ExpectedTypeFinder.java

```diff
@@ -0,0 +1,1138 @@
+                return Optional.empty();
+            }
+
+            return SymbolUtils.getTypeDescriptor(symbol.get());
+        }
+
+        return getExpectedType(bLangNode);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(AnnotationNode node) {
+        Node annotationRef = node.annotReference();
+        Optional<Symbol> annotationSymbol;
+        Predicate<Symbol> predicate = symbol -> symbol.getName().isPresent() &&
+                symbol.kind() == io.ballerina.compiler.api.symbols.SymbolKind.ANNOTATION;
+
+        if (annotationRef.kind() == SyntaxKind.QUALIFIED_NAME_REFERENCE) {
+            QualifiedNameReferenceNode qNameRef = (QualifiedNameReferenceNode) annotationRef;
+            Predicate<Symbol> qNamePredicate =
+                    predicate.and(symbol -> symbol.getName().get().equals(qNameRef.identifier().text()));
+            annotationSymbol = this.getTypeFromQNameReference(qNameRef, qNamePredicate);
+        } else if (annotationRef.kind() == SyntaxKind.SIMPLE_NAME_REFERENCE) {
+            annotationSymbol = this.getSymbolByName(((SimpleNameReferenceNode) annotationRef).name().text(), predicate);
+        } else {
+            return Optional.empty();
+        }
+
+        if (annotationSymbol.isEmpty()) {
+            return Optional.empty();
+        }
+
+        return ((AnnotationSymbol) annotationSymbol.get()).typeDescriptor();
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(SpecificFieldNode node) {
+        Optional<TypeSymbol> elementType = Optional.empty();
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
+        }
+
+        if (bLangNode instanceof BLangRecordLiteral.BLangRecordVarNameField &&
+                ((BLangRecordLiteral.BLangRecordVarNameField) bLangNode).symbol != null &&
+                node.parent().kind() == SyntaxKind.MAPPING_CONSTRUCTOR) {
+                BLangExpression bLangExpression = ((BLangRecordLiteral.BLangRecordVarNameField) bLangNode)
+                        .impConversionExpr;
+                if (bLangExpression != null && bLangExpression.getBType().getKind() == TypeKind.ANY) {
+                        BType bType = bLangExpression.getBType();
+                        return getTypeFromBType(bType);
+                }
+        }
+
+        if (bLangNode instanceof BLangRecordLiteral.BLangRecordVarNameField) {
+            elementType = getTypeFromBType(((BLangRecordLiteral.BLangRecordVarNameField) bLangNode).symbol.getType());
+        }
+
+        if (elementType.isPresent()) {
+            return elementType;
+        }
+
+        return getExpectedType(bLangNode);
+    }
+
+    @Override
+    public Optional<TypeSymbol> transform(ObjectFieldNode node) {
+        BLangNode bLangNode = nodeFinder.lookup(this.bLangCompilationUnit, node.lineRange());
+        if (bLangNode == null) {
+            return Optional.empty();
```

Comment:
This `null` check seems to be unnecessary. We have handled `null` in `getExpectedType(bLangNode);`. Please check other places as well